junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java

```diff
@@ -69,6 +70,27 @@
 		return testEngines;
 	}
 
+	/**
+	 * @return {@code true} if the graph does _not_ contain a cycle; else {@code false}.
+	 */
+	static boolean isAcyclic(TestDescriptor root) {
+		Set<UniqueId> visited = new HashSet<>();
+		visited.add(root.getUniqueId());
+		Stack<TestDescriptor> stack = new Stack<>();
+		stack.push(root);
+		while (!stack.isEmpty()) {
+			for (TestDescriptor child : stack.pop().getChildren()) {
+				if (!visited.add(child.getUniqueId())) {
+					return false; // id already known: cycle detected!
+				}
+				if (child.isContainer()) {
+					stack.add(child);
+				}
+			}
+		}
+		return true;
+	}
```

Comment:
@sormuras I'm not clear on what algorithm `isAcyclic()` uses to check for cycles. Can you place a comment towards the top of the body so that it's clear, so something like the following?

```java
static boolean isAcyclic() {
    // <name-of-algorithm-here>
    . . .
}
```

Reply:
Hm, never invented a name for some lines of code before. _stack-based-traversing-with-bread-crumb-checks_?

Reply:
Haha!

I was hoping that your algorithm had an official name like _Kahn's algorithm_ for topological sorting or _Tarjan's algorithm_ for finding strongly-connected components, as I understand from these [StackOverflow](https://stackoverflow.com/a/261595/2252930) [answers](https://stackoverflow.com/a/261621/2252930) that those algorithms can be used to find cycles in graphs.

If alternatively your algorithm is home-grown and it was inspired by an existing graph traversal algorithms like _depth-first search_ or _breadth-first search_, then it would help clarify things (for me, at least) if that was noted down somewhere, along with what change(s) were made to the traversal algorithm to allow cycles to be detected.

But maybe I'm asking for _way_ too much here, in which case I'd be happy to slowly read through the code in my own time later today to figure out what it's doing. :)

Reply:
Well, we don't actually have a _directed graph_. Rather, we have a _tree_.

So we should probably change the title of the issue and PR to avoid confusion. 😉

Reply:
And that's a breadth-first traversal of the tree of test descriptors.

Reply:
> Well, we don't actually have a directed graph. Rather, we have a tree.

A _tree_ is still a _graph_.

> So we should probably change the title of the issue and PR to avoid confusion.

 No where "directed" is mentioned. ;)

Reply:
Sure, but I'd still refer to it as a _tree_ in this context. 😇

Reply:
> And that's a breadth-first traversal of the tree of test descriptors.

@sbrannen Cool, thank you for the clarification/confirmation! 😀👍

Reply:
Ah cool, rereading the code with the knowledge that it uses breadth-first search, I understand what it's doing now: it's using a `stack` and continuous calls to `stack.pop().getChildren()` to iterate over the entire tree in breath-first order, and it keeps track of `visited` nodes to ensure there are no cycles. Got it! 👍

I'd personally love it if a minor comment were placed at the top of the body like `// based on breath-first search`, but I won't insist. 🙂

Reply:
Well, if @sormuras describes it as "breath-first [sic]", I would find that totally amusing...

Take a deep _breath_ first... and then dive deep! 🤣

Reply:
Haha! *facepalm*

This is what I get for typing things with my phone. :P